ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/multiplechoice/MultipleChoiceTaskViewModel.kt

```diff
@@ -16,15 +16,108 @@
+import com.google.android.ground.model.submission.Value
+import com.google.android.ground.model.task.MultipleChoice
 import com.google.android.ground.model.task.Option
+import com.google.android.ground.model.task.Task
 import com.google.android.ground.ui.datacollection.tasks.AbstractTaskViewModel
 import javax.inject.Inject
 
 class MultipleChoiceTaskViewModel @Inject constructor(resources: Resources) :
   AbstractTaskViewModel(resources) {
 
-  fun updateResponse(options: List<Option>) {
-    setValue(fromList(task.multipleChoice, options.map(Option::id)))
+  private val backingItems: MutableLiveData<List<MultipleChoiceItem>> = MutableLiveData()
+  private val selectedIds: MutableList<String> = mutableListOf()
+  val itemsLivedata: LiveData<List<MultipleChoiceItem>> = backingItems
+
+  private val otherId: String = "OTHER_ID"
+  private var otherText: String = ""
+
+  override fun initialize(job: Job, task: Task, value: Value?) {
+    super.initialize(job, task, value)
+    updateMultipleChoiceItems()
   }
+
+  /* Reads the list of options in a multiple choice object and converts them to MultipleChoiceItems*/
+  private fun updateMultipleChoiceItems() {
+    val multipleChoice = checkNotNull(task.multipleChoice)
+    val isMultipleChoice = multipleChoice.cardinality == MultipleChoice.Cardinality.SELECT_MULTIPLE
+    val items: MutableList<MultipleChoiceItem> = mutableListOf()
+
+    items.addAll(
+      multipleChoice.options.map { option ->
+        MultipleChoiceItem(
+          option,
+          isOptionSelected(option),
+          if (isMultipleChoice) MultipleChoiceItem.Type.CHECKBOX else MultipleChoiceItem.Type.RADIO,
+        )
+      }
+    )
+
+    if (multipleChoice.hasOtherOption) {
+      val otherOption = Option(otherId, "", "Other")
+      items.add(
+        MultipleChoiceItem(
+          otherOption,
+          isOptionSelected(otherOption),
+          if (isMultipleChoice) MultipleChoiceItem.Type.CHECKBOX else MultipleChoiceItem.Type.RADIO,
+          true,
+        )
+      )
+    }
+    backingItems.value = items
+  }
+
+  private fun isOptionSelected(option: Option): Boolean {
+    return selectedIds.contains(option.id)
+  }
+
+  fun onOptionClicked(item: MultipleChoiceItem) {
+    val isSelected = !isOptionSelected(item.option)
+    if (item.type == MultipleChoiceItem.Type.RADIO) {
+      selectedIds.clear()
+      if (isSelected) selectedIds.add(item.option.id)
+    } else {
+      if (isSelected) {
+        selectedIds.add(item.option.id)
+      } else {
+        selectedIds.remove(item.option.id)
+      }
+    }
+    updateResponse()
+  }
+
+  fun updateResponse() {
+    setValue(
+      fromList(
+        task.multipleChoice,
+        selectedIds.map { if (it == otherId) formatOtherText(otherText) else it },
+      )
+    )
+    updateMultipleChoiceItems()
+  }
+
+  private fun formatOtherText(otherText: String): String {
+    return "[ $otherText ]"
+  }
```

Comment:
Is this formatting applied to other values as well? Or is it a product requirement for "other" type?

Reply:
This is done just for the other type. @gino-m can give more info.
But this is a hacky way to have android upload user answers without updating the backend to accept the new value. We are basically just making up an ID that contains the user created text.